### PR TITLE
Add initial oneOf schema code gen support

### PR DESCRIPTION
**Pull request by OwenGraves**
This uses shrubbery untagged unions to implement oneOf code gen. This adds better support for handling $refs inside the components/schemas section of the spec, allowing for parse errors on missing ref names.

  - This converts additionalProperties refs to the new approach used by oneOf, since previously a missing ref name in additionalProperties under components/schemas would be a compile instead of parse error
  - This does not fix nullable for inline arrays under the operations section (or implement them correctly for inline oneOf), though we could as part of this pr if deemed important enough
  - This does not implement openapi discriminators, which would use tagged unions
  - This does not implement inline objects as choices for `oneOf`
  - If we run into `anyOf` being used, it would probably be simple to have it just call out to the `oneOf` code (though hopefully no one ever uses `anyOf`...)

### CODE REVIEW

File: json-fleece-openapi3/src/Fleece/OpenApi3.hs [hunk 1]
```diff
@@ -398,99 +488,84 @@ mkInlineBodySchema raiseError schemaKey schemaMap schema =
...
+  (forall a. String -> CGU.CodeGen a) ->
+  T.Text ->
+  SchemaMap ->
+  OA.Schema ->
+  CGU.CodeGen SchemaTypeInfoWithDeps
+mkInlineArrayOneOfSchema raiseError schemaKey schemaMap schema =
+  case OA._schemaItems schema of
+    Just (OA.OpenApiItemsObject (OA.Ref ref)) -> do
+      pure $
+        -- We don't currently have a good way to get the nullability of the referenced schema
```
Comment (qxjit) [inline json-fleece-openapi3/src/Fleece/OpenApi3.hs:518]: This is probably an issues we could fix because it leaves things in state that will generate code that doesn't match the OpenAPI description in a very subtle and non-obvious way.

**Review by qxjit — CHANGES_REQUESTED** [review]

File: json-fleece-openapi3/src/Fleece/OpenApi3.hs (hunk 1 above)
Comment (OwenGraves) [reply json-fleece-openapi3/src/Fleece/OpenApi3.hs]: I added several tests and support for inline nullability, there are still some lingering edge cases (with other features we haven't fully covered yet like inline nested objects, or nested oneOf), but stuff like `nullable: true` with a `$ref` sibling doesn't even appear to be valid according to the spec, which I didn't realize at first.

File: json-fleece-openapi3/src/Fleece/OpenApi3.hs [hunk 2]
```diff
@@ -241,24 +242,24 @@ lookupRequestBody operationKey operation =
       pure Nothing
 
 data SchemaTypeInfoWithDeps = SchemaTypeInfoWithDeps
-  { schemaTypeInfoDependent :: CGU.SchemaTypeInfo
+  { schemaTypeInfoDependent :: CGU.SchemaTypeInfoOrRef
```
[the remark this answers is not in the document]
Comment (OwenGraves) [reply json-fleece-openapi3/src/Fleece/OpenApi3.hs:245]: I agree there could probably be some further simplification in the abstractions here around `SchemaTypeInfoWithDeps` (actually `SchemaTypeInfoOrRef` was one initial attempt in that direction). The tricky thing is we have refs which need to be resolved later (at code generation time, once we have all the top level names, since refs under components could refer to other refs in components), but we also have a lot of code (especially the operations section) that directly produces `SchemaTypeInfo` (ex: `lookupRequestBodySchema`). I think it might be feasible to move all the `SchemaTypeInfo` resolution into `generateCodeGenDataFormat` by expanding `resolveRefTypeInfo` to work on more than just refs for instance, but this would be a pretty significant change that I'd want to discuss through first. I also think it may be reasonable to wait and do this refactoring at a later stage, as implementing some of the unsupported features (like nested inline objects) might change how we approach the abstractions.

**Review by qxjit — APPROVED** [review]
